feat(core): allow for injector to be specified when creating an embedded view

crisbeto opened the pull request:
Adds support for passing in an optional injector when creating an embedded view through `ViewContainerRef.createEmbeddedView` and `TemplateRef.createEmbeddedView`. The injector allows for the DI behavior to be customized within the specific template.

This is a second stab at the changes in #44666. The difference this time is that the new injector acts as a node injector, rather than a module injector.

Fixes #14935.

crisbeto commented on packages/core/src/render3/di.ts:
```diff
@@ -406,100 +396,126 @@ export function getOrCreateInjectable<T>(
+ * @param notFoundValue The value to return when the injection flags is `InjectFlags.Optional`
+ * @returns the value from the injector, `null` when not found, or `notFoundValue` if provided
+ */
+function lookupTokenUsingNodeInjector<T>(
```
This function's code is the same as `getOrCreateInjectable` currently, feel free to ignore when reviewing.

crisbeto commented on packages/core/test/render3/perf/embedded_view_injector/index.ts:
```diff
@@ -0,0 +1,47 @@
+
+const noInjectorHarness = setupTestHarness(template, 1, 0, 1, {}, null, [noInjectorApp.ɵcmp]);
+const withInjectorHarness = setupTestHarness(template, 1, 0, 1, {}, null, [withInjectorApp.ɵcmp]);
+
```
There's something weird going on in this benchmark, because whichever case runs first ends up being quicker, even if it I give it identical components by not passing in an injector above. The difference can be observed by swapping the order of the `while` loops below.

cc @pkozlowski-opensource

alxhub replied:
That's super interesting. One hypothesis is that it could be a branch misprediction issue, where we train the branch predictor to expect one case (no embedded injector) and then violate that assumption in the next case (embedded injector present).

crisbeto replied:
It's worth noting that the same happens if I create two identical components that don't hit the new code path.

dylhunn reviewed (approved):
reviewed-for: fw-core, fw-common, public-api

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api, fw-core, fw-common, size-tracking

alxhub reviewed (approved):
Reviewed-for: public-api

dylhunn reviewed (approved):
reviewed-for: public-api, fw-core, fw-common, size-tracking

jessicajaniuk commented:
This PR was merged into the repository by commit 94c949a60ad0de1f5385ec97f2c1933cb1d55f22.

dylhunn commented:
We are going to roll this back, we think it caused a failure in sync cl/431510120.

crisbeto committed:
feat(core): allow for injector to be specified when creating an embedded view

Adds support for passing in an optional injector when creating an embedded view through `ViewContainerRef.createEmbeddedView` and `TemplateRef.createEmbeddedView`. The injector allows for the DI behavior to be customized within the specific template.

This is a second stab at the changes in #44666. The difference this time is that the new injector acts as a node injector, rather than a module injector.

Fixes #14935.

crisbeto commented:
[Passing TGP](https://fusion2.corp.google.com/presubmit/tap/431857780/OCL:431857780:BASE:431872883:1646210501451:634d5aa7) after a fix yesterday where one of the bitmaps was missing a bit. Unblocking.

jessicajaniuk commented:
This PR was merged into the repository by commit 69018c9f42d14c1c7aa8271b406d6d247de1c564.
